src/main/scala/mesosphere/mesos/ResourceMatcher.scala

```diff
@@ -15,66 +16,88 @@ import scala.collection.mutable
+    lazy val hostPorts: Seq[Int] = portsMatch.hostPorts
+
+    def scalarMatch(name: String): Option[ScalarMatch] = scalarMatches.find(_.resourceName == name)
+
+    def resources: Iterable[org.apache.mesos.Protos.Resource] =
+      scalarMatches.flatMap(_.consumedResources) ++ portsMatch.resources
+  }
+
+  /**
+    * Restricts which resources are considered for matching.
+    *
+    * Disk resources are always discarded, since we do not want to match them by
+    * accident.
+    *
+    * @param acceptedRoles contains all Mesos resource roles that are accepted
+    * @param reserved if reserved is true, only resources with a ReservationInfo
+    *                 are considered. If reserved is false, only resources without
+    *                 a ReservationInfo are considered.
+    */
+  case class ResourceSelector(acceptedRoles: Set[String], reserved: Boolean) {
+    def apply(resource: Protos.Resource): Boolean = {
+      // resources with disks are matched by the VolumeMatcher or not at all
+      val noAssociatedDisk = !resource.hasDisk
+
+      noAssociatedDisk && acceptedRoles(resource.getRole) && resource.hasReservation == reserved
+    }
 
     override def toString: String = {
-      if (matches) {
-        s"SATISFIED ($requiredValue <= $offeredValue)"
-      }
-      else {
-        s"NOT SATISFIED ($requiredValue > $offeredValue)"
-      }
+      val reservedString = if (reserved) "RESERVED" else "unreserved"
+      val rolesString = acceptedRoles.mkString(", ")
+      s"Considering $reservedString resources with roles {$rolesString}"
     }
   }
 
-  private[this] val log = LoggerFactory.getLogger(getClass)
-
-  case class ResourceMatch(cpuRole: Role, memRole: Role, diskRole: Role, ports: Seq[RangesResource])
+  object ResourceSelector {
+    /** Match unreserved resources of role '*' (the default role) */
+    def wildcard: ResourceSelector = ResourceSelector(Set("*"), reserved = false)
+  }
 
-  //scalastyle:off method.length
+  /**
+    * Checks whether the given offer contains enough resources to launch a task of the given app
+    * or to make a reservation for a task.
+    *
+    * If a task uses local volumes, this method is typically called twice for every launch. Once
+    * for the reservation on UNRESERVED resources and once for every (re-)launch on RESERVED resources.
+    *
+    * If matching on RESERVED resources as specified by the ResourceSelector, resources for volumes
+    * have to matched separately (e.g. by the [[PersistentVolumeMatcher]]). If matching on UNRESERVED
+    * resources, the disk resources for the local volumes are included since they must be part of
+    * the reservation.
+    */
   def matchResources(offer: Offer, app: AppDefinition, runningTasks: => Iterable[Task],
-                     acceptedResourceRoles: Set[String] = Set("*")): Option[ResourceMatch] = {
+                     selector: ResourceSelector = ResourceSelector.wildcard): Option[ResourceMatch] = {
```

Comment:
Is the default value a leftover? Production code always explicitly sets the ResourceSelector and I think it would be better to not assume a default.